### PR TITLE
2024 robot template

### PR DESCRIPTION
**Pull request by vichik123**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2,7 +2,7 @@ import edu.wpi.first.gradlerio.deploy.roborio.RoboRIO;
 
 plugins {
     id "java"
-    id "edu.wpi.first.GradleRIO" version "2023.1.1"
+    id "edu.wpi.first.GradleRIO" version "2023.4.3"
```
Comment (katzuv) [inline build.gradle]: Bump to 2024 beta version.

File: vendordeps/AdvantageKit.json [hunk 2]
```diff
@@ -1,39 +1,39 @@
 {
   "fileName": "AdvantageKit.json",
   "name": "AdvantageKit",
-  "version": "2.0.0-beta-3",
+  "version": "2.2.4",
```
Comment (katzuv) [inline vendordeps/AdvantageKit.json]: You should bump to the 2024 beta version.

File: networktables.json
Comment (katzuv) [inline networktables.json:1]: What is the purpose of this file?

File: simgui-ds.json
Comment (katzuv) [inline simgui-ds.json:1]: What program is using this file and the `simgui.json` file?

File: src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package frc.robot.subsystems.example.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.example.ExampleSubsystem;
+
+public class ExampleCommand extends CommandBase {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java]: Is this file really needed?

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     * @return The single instance of ExampleSubsystem.
+     */
+    public static ExampleSubsystem getINSTANCE() {
+        if (INSTANCE == null) {
+            if (Robot.isReal()) {
+                INSTANCE = new ExampleSubsystem(new ExampleSubsystemIOReal());
+            } else {
+                INSTANCE = new ExampleSubsystem(new ExampleSubsystemIOSim());
+            }
+        }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: Subsystems don't need to be aware of their "realness", instead, pass the IO as a parameter to this constructor.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    private ExampleSubsystem(ExampleSubsystemIO io) {
+        this.io = io;
+    }
+
+    /**
+     * Gets the single instance of ExampleSubsystem.
+     *
+     * @return The single instance of ExampleSubsystem.
+     */
+    public static ExampleSubsystem getINSTANCE() {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: Rename to `geInstance()`.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package frc.robot.subsystems.example;
+
+public class ExampleSubsystemConstants {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java:7]: Very happy to see subsystem-specific constants returning to the subsystem package! In recent years there were always many merge conflicts.

First and last year subsystem constants were in this form [was in 2019](https://github.com/Galaxia5987/robot-2019/blob/master/src/main/java/robot/subsystems/climb/Constants.java) (when I was team lead 😉)

File: src/test/java/frc/robot/common/utils/TestUtils.java
Comment (katzuv) [inline src/test/java/frc/robot/common/utils/TestUtils.java:1]: This file already exists in the common repository.

File: src/main/java/frc/robot/Robot.java [hunk 7]
```diff
@@ -34,22 +36,25 @@ public class Robot extends LoggedRobot {
...
-        autonomousCommand = robotContainer.getAutonomousCommand();
 
-        Logger.getInstance().recordMetadata("ProjectName", "Wcp-Swerve-2023"); // Set a metadata value
+        Logger.getInstance().recordMetadata("ProjectName", "Robot-template");
 
         if (isReal()) {
-            Logger.getInstance().addDataReceiver(new WPILOGWriter("/media/sda1/")); // Log to a USB stick
-            Logger.getInstance().addDataReceiver(new NT4Publisher()); // Publish data to NetworkTables
-            new PowerDistribution(1, PowerDistribution.ModuleType.kRev); // Enables power distribution logging
+            Logger.getInstance().addDataReceiver(new NT4Publisher());
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: Adapt code to AdvantageKit 2024.

File: src/main/java/frc/robot/Robot.java [hunk 8]
```diff
@@ -34,22 +36,25 @@ public class Robot extends LoggedRobot {
...
+            Logger.getInstance().addDataReceiver(new WPILOGWriter("home/lvuser"));
+
+            try (PowerDistribution pwr = new PowerDistribution(1, PowerDistribution.ModuleType.kRev)) {
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: Use `LoggedPowerDistribution ` instead. `LoggedDriverStation` and `LoggedSystemStats` exist too.

**Review by katzuv — CHANGES_REQUESTED** [review]
Good job!!!

File: src/main/java/frc/robot/Robot.java [hunk 9]
```diff
@@ -29,6 +32,8 @@ public class Robot extends LoggedRobot {
     private RobotContainer robotContainer;
     private Command autonomousCommand;
 
+    private final TalonFX motor = new TalonFX(1, "caniboy");
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: Find a more professional name for the CANivore. Also, I'd rename the motor to better reflect it's specifically connected to a CANivore.

File: src/main/java/frc/robot/Robot.java [hunk 10]
```diff
@@ -67,6 +74,8 @@ public void robotInit() {
     @Override
     public void robotPeriodic() {
         CommandScheduler.getInstance().run();
+
+        motor.setControl(new DutyCycleOut(1));
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: From the [Phoenix 6 documentation](https://pro.docs.ctr-electronics.com/en/latest/docs/api-reference/api-usage/control-requests.html#modifying-a-control-request):
> Java users should reuse control requests to prevent excessive invocation of the Garbage Collector.

You can create the request object as a class member and change the value as needed with [`withPosition()`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/controls/PositionVoltage.html#withPosition(double)) (this doesn't create a new object but changes the same one).

File: vendordeps/Phoenix6.json [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+  "frcYear": 2024,
+  "uuid": "e995de00-2c64-4df5-8831-c1441420ff19",
+  "mavenUrls": [
+    "https://maven.ctr-electronics.com/release/"
+  ],
+  "jsonUrl": "https://maven.ctr-electronics.com/release/com/ctre/phoenix6/latest/Phoenix6-frc2024-beta-latest.json",
+  "conflictsWith": [
+    {
+      "uuid": "3fcf3402-e646-4fa6-971e-18afe8173b1a",
+      "errorMessage": "The combined Phoenix-6-And-5 vendordep is no longer supported. Please remove the vendordep and instead add both the latest Phoenix 6 vendordep and Phoenix 5 vendordep.",
```
Comment (katzuv) [inline vendordeps/Phoenix6.json:14]: Do we still use Phoenix 5 devices? If so, the docs still state the `Phoenix6And5` vendordep still exists: https://pro.docs.ctr-electronics.com/en/latest/docs/installation/installation.html#online
I'd try removing `Phoenix5.json` and `Phoenix6.json`, adding the combined JSON, and see whether it works, or this error message is real.

**Review by katzuv — CHANGES_REQUESTED** [review]
Updating versions is coollllll

File: networktables.json
Comment (vichik123) [reply networktables.json:1]: Not sure, I think it's created when you open the shuffleboard.

File: simgui-ds.json
Comment (vichik123) [reply simgui-ds.json:1]: The simulation program, I could put these and the networktables.json in the gitignore.

File: src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java (hunk 3 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java]: Yes, and it's a good reference.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java (hunk 4 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: I don't see why it shouldn't be aware of its realness.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java (hunk 5 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: Did it.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java (hunk 6 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java:7]: These existed in 2023 too, there was just an additional file for constants for all subsystems.

File: src/main/java/frc/robot/Robot.java (hunk 7 above)
Comment (vichik123) [reply src/main/java/frc/robot/Robot.java]: Did it.

File: src/main/java/frc/robot/Robot.java (hunk 9 above)
Comment (vichik123) [reply src/main/java/frc/robot/Robot.java]: I did this as a temporary experiment, I don't know why I pushed it.

File: src/main/java/frc/robot/Robot.java (hunk 10 above)
Comment (vichik123) [reply src/main/java/frc/robot/Robot.java]: I'll change it.

File: vendordeps/Phoenix6.json (hunk 11 above)
Comment (vichik123) [reply vendordeps/Phoenix6.json:14]: I don't want to use phoenix 5, but I want to leave it just in case. Why do you want me to use the merged json?

File: src/main/java/frc/robot/Robot.java (hunk 10 above)
Comment (vichik123) [reply src/main/java/frc/robot/Robot.java]: Ah hold on, I deleted this anyway.

File: src/main/java/frc/robot/Robot.java (hunk 9 above)
Comment (katzuv) [reply src/main/java/frc/robot/Robot.java]: No worries. Now you can use interactive rebasing to delete/edit the commit that introduced this change 🙂

File: networktables.json
Comment (katzuv) [reply networktables.json:1]: Is it safe to add to `.gitignore`?

File: simgui-ds.json
Comment (katzuv) [reply simgui-ds.json:1]: Sure, go ahead.

File: src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java (hunk 3 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/example/commands/ExampleCommand.java]: I differ, but your choice.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java (hunk 4 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: It breaks the principle that subsystems are isolated from their implementations. It also means that you're limited to two implementations -- how do you plan to add a NEO IO for the swerve in Robotinyo that way?

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java (hunk 5 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: Nice. Next time, don't forget to include the commit's SHA.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java (hunk 6 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/example/ExampleSubsystemConstants.java:7]: I did not know that. Good to know 🙂
> there was just an additional file for constants for all subsystems.

There is one in this template too, as should be.

File: vendordeps/Phoenix6.json (hunk 11 above)
Comment (katzuv) [reply vendordeps/Phoenix6.json:14]: Easier to maintain and update one file instead of two.

Comment (katzuv) [reply vendordeps/Phoenix6.json:14]: I changed my mind. IMO it's best to delete the Phoenix 5 file, right now there's a risk of importing the same class from Phoenix 5 instead of Phoenix 6. Removing this file will eliminate this risk, we can always add it back if needed.

File: src/main/java/frc/robot/Robot.java (hunk 9 above)
Comment (katzuv) [reply src/main/java/frc/robot/Robot.java]: Will be changed in common instead.

**Review by katzuv — APPROVED** [review]

**Comment by katzuv** [comment]
Can we merge this?

**Comment by katzuv** [comment]
Can we merge this?

**Review by katzuv — APPROVED** [review]
🎵
